ops/testing.py

```diff
@@ -1759,6 +1789,104 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class _MockStorage:
+    def __init__(self, location, src):
+        self._src = src
+        os.makedirs(src, exist_ok=True)
+        self._location = location
+
+    def contains(self, path):
+        return str(path).startswith(self._location)
+
+    def _srcpath(self, path):
+        suffix = path[len(self._location):]
+        if suffix[0] == '/':
+            suffix = suffix[1:]
+        srcpath = os.path.join(self._src, suffix)
+        return srcpath
+
+    def create_dir(self, path: str, make_parents: bool = False, **kwargs) -> '_Directory':
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        dirname = os.path.dirname(srcpath)
+        if not os.path.exists(dirname):
+            if not make_parents:
+                raise RuntimeError('no such directory {}'.format(dirname))
+            os.makedirs(dirname)
+        os.makedirs(srcpath)
+        return _Directory(self, path, **kwargs)
+
+    def create_file(
+            self,
+            path: str,
+            data: typing.Union[bytes, str, typing.BinaryIO, typing.TextIO],
+            encoding: typing.Optional[str] = 'utf-8',
+            make_dirs: bool = False,
+            **kwargs
+    ) -> '_File':
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        dirname = os.path.dirname(srcpath)
+        if not os.path.exists(dirname):
+            if not make_dirs:
+                raise RuntimeError('no such directory {}'.format(dirname))
+            os.makedirs(dirname)
+
+        if not isinstance(data, bytes):
+            data = data.encode(encoding=encoding)
+
+        with open(srcpath, 'wb') as f:
+            f.write(data)
+
+        return _File(pathlib.PurePosixPath(path), data, encoding=encoding, **kwargs)
+
+    def list_dir(self, path) -> typing.List['_File']:
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        results = []
+        for fname in os.listdir(srcpath):
+            fpath = os.path.join(srcpath, fname)
+            mountpath = os.path.join(path, fname)
```

Comment:
I believe you can do `list_dir` for subdirectories as well. eg:
```
location = '/foo'
storage.mkdir('bar')
storage.mkdir('bar/baz')
storage.list_dir('bar/baz')
```

I don't see much in the way of direct testing of _MockStorage. This is probably something that should be covered to make sure we work correctly. (My second look at it says that it might, but I thought it would probably be good to verify)

I also think this code gets slightly cleaner with `pathlib`, so would recommend it.
(Note that locations can be used with pathlib.PurePosixPath, but src probably needs to use pathlib.Path)

Reply:
Pathlib conversion done.  I'll make sure tests get well rounded out too.

Reply:
Alright - I've converted the testingfilesystem tests to a mixin that is now applied to both the TestingFilesystem and the TestingStorageMount classes (and fixed a few inconsistencies the tests caught).